zilliqa/src/api/eth.rs

```diff
@@ -1039,10 +1039,114 @@ fn blob_base_fee(_params: Params, _node: &Arc<Mutex<Node>>) -> Result<()> {
     Err(anyhow!("API method eth_blobBaseFee is not implemented yet"))
 }
 
+/// Calculates the gas rewards for a block based on specified percentiles.
+///
+/// # Parameters
+/// - `block`: The block for which the rewards are being calculated.
+/// - `reward_percentiles`: A slice of percentiles (0-100) at which to calculate the rewards.
+/// - `transactions_sorted_gas_used`: A vector of gas used by transactions, sorted in ascending order.
+///
+/// # Returns
+/// A vector of rewards corresponding to the specified percentiles.
+fn calculate_rewards(
+    block_gas_used: u64,
+    reward_percentiles: &[f64],
+    sorted_gas_prices: &[u128],
+) -> Vec<f64> {
+    let target_gas_values: Vec<f64> = reward_percentiles
+        .iter()
+        .map(|&percentile| percentile / 100.0 * (block_gas_used as f64))
+        .collect();
+
+    let mut rewards = vec![0.0; reward_percentiles.len()];
+    let mut sum_gas = 0.0;
+    let mut percentile_index = 0;
+
+    for gas_used in sorted_gas_prices.iter().map(|&g| g as f64) {
+        sum_gas += gas_used;
+
+        while percentile_index < reward_percentiles.len()
+            && sum_gas >= target_gas_values[percentile_index]
+        {
+            rewards[percentile_index] = gas_used;
+            percentile_index += 1;
+        }
+
+        if percentile_index >= reward_percentiles.len() {
+            break;
+        }
+    }
+
+    for i in percentile_index..reward_percentiles.len() {
+        rewards[i] = sorted_gas_prices.last().copied().unwrap_or(0) as f64;
+    }
+
+    rewards
+}
+
 /// eth_feeHistory
 /// Returns the collection of historical gas information
-fn fee_history(_params: Params, _node: &Arc<Mutex<Node>>) -> Result<()> {
-    Err(anyhow!("API method eth_feeHistory is not implemented yet"))
+fn fee_history(params: Params, node: &Arc<Mutex<Node>>) -> Result<eth::FeeHistory> {
+    let mut params = params.sequence();
+    let block_count: u64 = params.next()?;
+    let newest_block: BlockNumberOrTag = params.next()?;
+    let reward_percentiles: Option<Vec<f64>> = params.optional_next()?;
```

Comment:
```suggestion
    let reward_percentiles: Vec<f64> = params.optional_next()?.unwrap_or_default();
```
to save the extra branching later?

Reply:
I believe we should avoid including `reward` in the response if `reward_percentiles` is absent. Doesn't this imply that we need at least one branch?

Reply:
You are right, sorry